wlauto/devices/android/gem5/__init__.py

```diff
@@ -0,0 +1,820 @@
+        self.gem5_shell("busybox cp /mnt/obb/{} {}".format(filename, dest))
+        self.gem5_shell("busybox sync")
+        self.gem5_shell("ls -al {}".format(dest))
+        self.gem5_shell("ls -al /mnt/obb/")
+        self.logger.debug("Push complete.")
+
+    def pull_file(self, source, dest, as_root=False):  # pylint: disable=W0221
+        """
+        Pull a file from the gem5 device using m5 writefile
+
+        First, we check the extension of the file to be copied. If the file ends
+        in .gz, then gem5 wrongly assumes that it should create a gzipped output
+        stream, which results in a gem5 error. Therefore, we rename the file on
+        the local device prior to the writefile command when required. Next, the
+        file is copied to the local directory within the guest as the m5
+        writefile command assumes that the file is local. The file is then
+        written out to the host system using writefile, prior to being moved to
+        the destination on the host.
+        """
+        filename = source.split('/')[-1]
+        self.logger.debug("Pulling {} from device.".format(filename))
+
+        # gem5 assumes that files ending in .gz are gzip-compressed. We need to
+        # work around this, else gem5 panics on us. Rename the file for use in
+        # gem5
+        if filename[-3:] == '.gz':
+            filename += '.fugem5'
+
+        self.logger.debug("pull_file {} {}".format(source, filename))
+        # We don't check the exit code here because it is non-zero if the source
+        # and destination are the same. The ls below will cause an error if the
+        # file was not where we expected it to be.
+        self.gem5_shell("busybox cp {} {}".format(source, filename), check_exit_code=False)
+        self.gem5_shell("busybox sync")
+        self.gem5_shell("ls -la {}".format(filename))
+        self.logger.debug('Finished the copy in the simulator')
+        self.gem5_util("writefile {}".format(filename))
+
+        if 'cpu' not in filename:
+            while not os.path.exists(os.path.join(self.gem5outdir, filename)):
+                time.sleep(1)
+
+        # Perform the local move
+        shutil.move(os.path.join(self.gem5outdir, filename), dest)
+        self.logger.debug("Pull complete.")
+
+    def delete_file(self, filepath, as_root=False):  # pylint: disable=W0221
+        """ Delete a file on the device """
+        self._check_ready()
+        self.gem5_shell("rm '{}'".format(filepath))
+
+    def file_exists(self, filepath):
+        """ Check if a file exists """
+        self._check_ready()
+        output = self.gem5_shell('if [ -e \'{}\' ]; then echo 1; else echo 0; fi'.format(filepath))
+        try:
+            if int(output):
+                return True
+            else:
+                return False
+        except ValueError:
+            if output:
+                return True
+            else:
+                return False
```

Comment:
Not sure I understand the above logic. If `output` is not an integer, then as long as it is not empty, assume the file exists?